### DISCUSSION TITLE
Apps and watchfaces will not run in the simulator (Ubuntu 20.04)

### DISCUSSION THREAD

**Discussion by BeanTosser** [zepp-health/zeppos-forum#29] [Q&A]
I installed the Zepp simulator in Ubuntu. I am able to get it to run not just the simulator app but the actual watch simulation - by pressing the "simulator" button in the app - without issue. However, I am not able to actually RUN any watch apps in the simulator.

I use "zeus dev" in the root of a watch app project directory, and it successfully adds the app the the list in the simulator. When I click to open the app within the simulator, however, nothing changes in the actual simulation. It just continues to show the same stock watch face, and the app list remains empty except for the watch settings gear/cog icon.

I've noticed a few errors in the developer console as well as the terminal that I opened the simulator with.

In the devtools I get:
`WebSocket connection to 'ws://192.168.166.188:7833/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
`

In the terminal:
`(node:7802) electron: Failed to load URL: http://zepp-os-staging.zepp.com/frameworks/v1.0.1/side-service.html with error: ERR_NAME_NOT_RESOLVED
`

I find that second one interesting as zepp-os-staging.zepp.com isn't a valid/existing domain if you type it into a URL bar. Is the linux version of the simulator just not up-to-date and referencing a resource at a now obsolete url?

**Comment by mrcoon** [comment]
@BeanTosser  This seems to be a failure to establish communication with the watch emulator, did you run the  `cd /opt/simulator/resources/firmware/ && sudo ./setup_for_linux.sh` command?
For more information, please refer to:  https://docs.zepp.com/docs/guides/tools/simulator/setup/#linux-version-installation

**Reply by BeanTosser** [reply]
Yes, I did run this command when initially installing - and I also just ran it again just to be sure. Not surprisingly, I just got a bunch of messages like the following:

`Reading package lists... Done
Building dependency tree       
Reading state information... Done
libfdt-dev is already the newest version (1.5.1-1).
libglib2.0-dev is already the newest version (2.64.6-1~ubuntu20.04.4).
libpixman-1-dev is already the newest version (0.38.4-0ubuntu2.1).
zlib1g-dev is already the newest version (1:1.2.11.dfsg-2ubuntu1.5).
0 upgraded, 0 newly installed, 0 to remove and 2 not upgraded.
`

I tested the simulator again just in case, but not luck.

A couple things I probably should have mentioned at the start:

- Initially I couldn't get the simulator to pop up the watch simulation at all after following the documentation's installation instructions exactly. I don't recall how exactly I found out I needed to do it, but through some process of digging around the internet I encountered the idea that I might need to install Qemu manually myself - something never mentioned in the Zepp documentation for the simulator setup. Again, I don't recall exactly what commands I typed to install qemu, but once I did, I was able to at least get a watch simulation to run - I just can't get any apps or watchfaces to push to it.
- When opening the simulator in the terminal, the following errors pop up:
`[8018:1227/090618.258990:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
[8018:1227/090618.282104:ERROR:gbm_wrapper.cc(275)] Failed to export buffer to dma_buf: No such file or directory (2)
`
That second one repeats at least ten times. I'm not sure _exactly_ what they mean, but I gather from some google research that they're related to chrome browser interactions with the GPU. Clearly chrome is not directly involved here (I don't even have chrome installed on my machine), but my best guess as to why chrome errors would pop up is that parts of the simulator are built on chrome's open-source code - probably the devtools windows at the very least. 

I wouldn't expect GPU errors to relate to the watch apps not running, but I could be mistaken.

**Comment by jiaolaibin** [comment]
Run emulator. Enter the ifconfig command to see if tap0 appears on the local network, as follows:
![image](https://user-images.githubusercontent.com/108327182/209790659-08cad4d6-e5b9-4d65-9c09-0955278a4b2e.png)
If it doesn't work, try "sudo apt-get install -y uml-utilities bridge-utils" to fix it.

**Comment by jiaolaibin** [comment]
You can also try ping [192.168.166.188 ](http://192.168.166.188/)to see if the local network works.

**Comment by BeanTosser** [comment]
IF config produced the expected/successful result.

Pinging that IP did not... I got:
`From 192.168.166.1 icmp_seq=19 Destination Port Unreachable
ping: sendmsg: Operation not permitted
`

These are both moot points, however... It finally occurred to me that I should try disabling my VPN since this clearly seemed to be some sort of network issue. I did so and, what do you know,  I have the Hello World app running perfectly in the simulator.

Apologies for wasting your time with a stupid oversight on my part. On the other hand, hopefully others who run into the same issue will run into this and have a quick, easy solution.
